File: src/AnalyticsService/Service/Queries/Handlers/GetStatsForPeriodQueryHandler.cs
using System.Data;
using AnalyticsService.Database.Api;
using AnalyticsService.Service.Model;
using MediatR;

namespace AnalyticsService.Service.Queries.Handlers;

/// <summary>
/// A handler class for obtaining general app stats for a period.
/// </summary>
internal sealed class GetStatsForPeriodQueryHandler : IRequestHandler<GetStatsForPeriodQuery, GeneralBatchStat>
{
    private readonly IDbConnection _dbConnection;

    private readonly IBatchStatRepository _batchStatRepository;

    public GetStatsForPeriodQueryHandler(IDbConnection conn, IBatchStatRepository repo)
    {
        _dbConnection = conn;
        _batchStatRepository = repo;
    }

    /// <inheritdoc/>
    public async Task<GeneralBatchStat> Handle(GetStatsForPeriodQuery request, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return GeneralBatchStat.Default;
        }
        var stats = (await _batchStatRepository
            .GetBatchStats(_dbConnection, request.StartDate, request.Period))
            .ToList();
        if (stats.Count == 0)
        {
            return GeneralBatchStat.Default;
        }

        var successRate = CalculateSuccessRate(
            stats.Count,
            stats.Count(i => i.IsSuccess())
        );
        return new GeneralBatchStat(
            (int)stats.Average(s => s.NumberOfDocuments),
            successRate,
            TimeSpan.FromTicks((long)stats.Average(d => d.RunTime.Ticks))
        );
    }

    /// <summary>
    /// Method for calculating a success rate of document batches.
    /// </summary>
    private static float CalculateSuccessRate(int all, int successful)
        => all / (float)successful;
}

File: src/AnalyticsService/Service/Queries/Handlers/GetStatsForPeriodQueryHandler.fs
namespace AnalyticsService.Service.Queries.Handlers

open System
open System.Data
open AnalyticsService.Service.Commands
open AnalyticsService.Service.Model
open AnalyticsService.Database.Repositories
open MediatR

/// A query for obtaining generic stats for a specific time period.
[<Sealed>]
type internal GetStatsForPeriodQueryHandler(
    conn: IDbConnection,
    batchRepo: IBatchStatRepository
) =
    let calculateSuccessRate all successful = all / float(successful)

    interface IRequestHandler<GetStatsForPeriodQuery, GeneralBatchStats> with
        member this.Handle(request, cancellationToken) =
            task {
                if not(cancellationToken.IsCancellationRequested) then
                    let stats =
                        batchRepo.GetBatchStatsForPeriod
                            request.WorkflowId
                            request.StartDate
                            request.Period
                            conn
                        |> Async.AwaitTask
                        |> Async.RunSynchronously
                        |> Seq.toList

                    if stats.Length = 0 then
                        return GeneralBatchStats.Default
                    else
                        return {
                            AvgNumberOfDocs =
                                stats
                                |> List.averageBy(fun i -> float i.NumberOfDocuments)
                                |> Math.Round
                                |> int
                            SuccessRate =
                                stats
                                |> List.filter(fun i -> i.IsSuccess)
                                |> List.length
                                |> calculateSuccessRate stats.Length
                            AvgDuration =
                                stats
                                |> List.averageBy(fun i -> float i.RunTime.Ticks)
                                |> int64
                                |> TimeSpan.FromTicks
                        }
                else
                    return GeneralBatchStats.Default
            }